pupil_src/shared_modules/video_capture/ndsi_backend.py

```diff
@@ -501,22 +501,20 @@ def auto_activate_source(self):
 
         src_sel, src_sel_labels = self.source_selection_list()
         if len(src_sel) <= 1 and src_sel[0] is None:
-            logger.warning("No camera was found on the remote host.")
+            logger.warning("No default device is available on the remote host.")
             return
 
         cam_ids = self.cam_selection_lut[self.g_pool.process]
 
         for cam_id in cam_ids:
-            try:
-                source_id = next(
-                    src_sel[i] for i, lab in enumerate(src_sel_labels) if cam_id in lab
-                )
-            except StopIteration:
-                source_id = None
-                logger.warning("No camera was found with ID: {}".format(cam_id))
+            for i, label in enumerate(src_sel_labels):
+                if cam_id in label:
+                    source_id = src_sel[i]
+                    self.activate(source_id)
+                    break
             else:
```

Comment:
Sorry for not being clear. :-/ I think the `next()` call is still the best way to find a match. What I was referring to was to move the log message to the outer `for cam_id in cam_ids` loop. This way the user does not get a warning for each condition that was not met but only one in case no condition was met at all.

Reply:
I hope I get it right this time. Both StopIteration and for/else are fairly new to me.

Reply:
Looks good! 

- A `StopIteration` is raised if a generator `return`s. See this tutorial on  https://realpython.com/introduction-to-python-generators/
- Related to `for-else`: https://realpython.com/quizzes/python-while-loop This quiz starts simple but is actually a bit tricky ;)